Feat: /welcome Supporters

Bouncey opened the pull request:
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.

Points to note:

1. I moved the label outside of the progress bar because at lower percentages, the label truncates relative to the `active` progress width, rendering it incomplete. 
2. `activeDonations` only tracks active donations in our db (an active donation is defined currently as a donation record without an `endDate`). This count is inaccurate due to it only holding records from the revamped donation page and the donation model. The count does not include paypal donations. I plan to write a donation migration script to add/remove donations so they reflect the state of stripe and paypal. This will then allow us to correctly manage donations through a UI.
3. The `Click here to become a Supporter` button links to `https://donate.freecodecamp.org`. A future PR can change this behaviour when we port over the new `/donate` page.

Non-Donator View:
![image](https://user-images.githubusercontent.com/18572518/49228619-aa97c900-f3e3-11e8-9461-f987175e7685.png)

Donator View:
![image](https://user-images.githubusercontent.com/18572518/49228687-ce5b0f00-f3e3-11e8-820f-af8c1efdd435.png)

Bouncey committed:
feat: Add supporter UI to /welcome

Bouncey committed:
fix: Move progress label out of the progress bar

Bouncey committed:
feat: Create and use dynamic active donation counts

QuincyLarson commented:
@Bouncey this looks great! We don't need the link to the survey or anything though - you can remove that completely.

I think for existing contributors we could also say (where the button would be for non supporters) "Thanks for being a supporter! Do you know anyone who's interested in technology? Encourage them to join the community as well."

Also, the number of supporters should have a comma in it.

Bouncey committed:
fix: Update Donators copy

Bouncey committed:
chore: Remove welcome link

Bouncey commented:
Removed the link and updated the copy

![image](https://user-images.githubusercontent.com/18572518/49235187-70cdbf00-f3f1-11e8-97c5-ad378eb172dc.png)

QuincyLarson commented:
@bouncey looks good. 

Could we get a comma in the supporter count?

Also, is this the real supporter count from Stripe? The values don't match between the progress bar and the "Join 4,000 supporters..." text.

Finally - is it possible to put the text inside the progress bar like in my example? I know it may look weird on smaller displays, but I think it will look better on desktop.

Bouncey commented:
Oh, I need to update the `Join **** Supporters` text. 

I plugged these numbers in myself to see how it would look.

The count, for now, will be generated from a db query:

> 2. `activeDonations` only tracks active donations in our db (an active donation is defined currently as a donation record without an `endDate`). This count is inaccurate due to it only holding records from the revamped donation page and the donation model. The count does not include paypal donations. I plan to write a donation migration script to add/remove donations so they reflect the state of stripe and paypal. This will then allow us to correctly manage donations through a UI.

QuincyLarson commented:
@bouncy will the database queries tax the system very much? For paypal donations, we currently have 1,703 paypal donors - could you set the value to be `variable + 1703` ?

I'm going to sleep but I've asked @raisedadead to help QA this so you can get this merged and deployed ASAP. Thanks for building this, and I'm excited to see the donation migration script.

QuincyLarson commented:
@Bouncey will the database queries tax the system very much? For paypal donations, we currently have 1,703 paypal donors - could you set the value to be `variable + 1703` ?

I'm going to sleep but I've asked @raisedadead to help QA this so you can get this merged and deployed ASAP. Thanks for building this, and I'm excited to see the donation migration script.

Bouncey committed:
feat: Comma separate active donations number

raisedadead commented:
> ...you can get this merged and deployed ASAP.

@QuincyLarson This will not be deployed until master is ready for production use.

Bouncey committed:
chore: Refactor server tests to use jest

Bouncey committed:
feat: Run server tests from project root test command

Bouncey commented:
I had planned to add some sort of caching mechanism, I just wanted some feedback on the overall look so far. 

> is it possible to put the text inside the progress bar like in my example? 

Yes, but we would struggle with contrast ratios if using white or black text:
![contrast comparisons](https://user-images.githubusercontent.com/18572518/49327547-fb7dfd80-f55a-11e8-8061-ae0cf292adfc.png)

Bouncey committed:
fix: Position label over the progress bar

Bouncey committed:
feat: InMemoryCache implementation

Bouncey committed:
feat: Use InMemoryCache to reducve load on db

Bouncey committed:
fix: Use env variable for paypal donations

QuincyLarson commented:
You should use line breaks inside the text. Otherwise people will just see a wall of text and not read it.

freeCodeCamp.org is a tiny nonprofit that's helping millions of people learn to code for free.
[line break]
Join 3,890 supporters. Your $5 / month donation will help keep tech education free and open.

And:
Thanks for being a supporter!
[line break]
Do you know...

The "N supporters out of 10,000 supporter goal" text should be centered beneath the progress bar. (note the slight wording change: "10,000 supporter goal")

Bouncey commented:
![copy-update](https://user-images.githubusercontent.com/18572518/49331849-63543880-f59b-11e8-9e8a-7f0cfd09e9e3.png)

Bouncey committed:
fix: Update copy and label position

Bouncey commented:
TODO:

- [ ] `inMemoryCache` error handling

Bouncey commented:
![image](https://user-images.githubusercontent.com/18572518/49338349-bff53900-f617-11e8-8e38-f6cd18f6b887.png)

QuincyLarson committed:
Remove words "as well" to prevent text wrapping

Bouncey committed:
feat: Add error handling for inMemoryCache

Bouncey committed:
fix: Use reportError in development

ValeraS commented on client/src/components/Supporters.js:
```diff
@@ -0,0 +1,76 @@
+          <ProgressBar max={10000} now={activeDonations} />
+          <div id='progress-label-wrapper'>
+            <span className='progress-label'>
+              {commaNumber(activeDonations)} supporters out of 10,000 supporter
```
Maybe `10,000` extract to a variable and use commaNumber for it.

ValeraS commented on client/src/components/Supporters.js:
```diff
@@ -0,0 +1,76 @@
+      </FullWidthRow>
+      <FullWidthRow>
+        <div id='supporter-progress-wrapper'>
+          <ProgressBar max={10000} now={activeDonations} />
```
What happens if `activeDonations` becomes greater `max`?

Bouncey replied:
Nice catch.

@QuincyLarson How would we want to handle something like

![image](https://user-images.githubusercontent.com/18572518/49367805-fd73c800-f6e4-11e8-8b87-1ddb6ff9fc9f.png)

raisedadead replied:
When we cross the goal: 

- Remove the progress bar.
- I think we can add a `🎉` and thank you message with the count, in a larger font.

Anyways that's all a good to have?

Bouncey replied:
Like this?

![image](https://user-images.githubusercontent.com/18572518/49371234-b7bbfd00-f6ee-11e8-828e-561db7765fb3.png)
![image](https://user-images.githubusercontent.com/18572518/49371274-d7ebbc00-f6ee-11e8-815f-cbc3943cf538.png)

QuincyLarson replied:
LGTM

Bouncey commented on api-server/server/utils/in-memory-cache.test.js:
```diff
@@ -0,0 +1,49 @@
+
+    it('can handle promises correctly', done => {
+      const cache = inMemoryCache(before);
+      cache.update(() => new Promise(resolve => resolve(after)));
```
This will not work. If you unwrap everything, it will look something like: 

```js
() => Promise.resolve(after))
  .then(() => {
    expect(cache.get()).toBe(after);
    done();	
  })
  .then(value => cache.set(cacheKey, value))
```

Which is an impossible test to pass

ValeraS replied:
`update` for promise returns `maybePromisedValue.then(value => cache.set(cacheKey, value))`. Why is `.then(value => cache.set(cacheKey, value))` last in your snippet?

Bouncey replied:
```js

update(fn) {
  const value = fn(); 
  /*
  fn = () => Promise.resolve(after).then(() => {
    expect(cache.get()).toBe(after);
    done();	
  })
  */

  if (isPromiseLike(value)) {
    value.then((value) => cache.set(cacheKey, value))

  /*
  which becomes -> 
  Promise.resolve(after)
    .then(() => {
      expect(cache.get()).toBe(after);
      done();	
    })
    .then((value) => cache.set(cacheKey, value))
  */
  }
}
```

Bouncey replied:
Ohhhhhhhhhhhh. Pesky brackets!

Fix pushed

Bouncey commented on client/src/utils/index.js:
```diff
@@ -20,3 +20,26 @@ export const getShortIdFromSlug = (slug = '') => {
   return maybeShortId.replace(/\/*$/, '');
 };
+
+export function commaNumber(num) {
```
Because sometimes we forget just how huge the javaScript language is and how it covers a lot of these things already...

Thanks for the reminder

Bouncey committed:
fix: Use built-ins to format numbers

Bouncey committed:
chore: Tidy up promise handling and testing

Bouncey committed:
fix: Fix up my broken test reformatting

Bouncey committed:
fix: Simplify async update testing

Bouncey committed:
fix: Handle goal reached 🎉

ValeraS commented on api-server/server/utils/in-memory-cache.js:
```diff
@@ -0,0 +1,34 @@
+      const value = cache.get(cacheKey);
+      return typeof value !== 'undefined' ? value : null;
+    },
+    async update(fn) {
```
1. Why do we need `fn` as parameter and not a value?
2. Why do not return promise always?

raisedadead committed:
chore: update lockfile
